content/blog/2017-11-10-ozunconf2017realtime.md

```diff
@@ -0,0 +1,93 @@
+
+When the call for nominations came around this year it was an easy decision. I don't have employer support to attend these things so I take time off work and pay my own way. This is my networking time, my development time, and my skill-building time. I wasn't sure what sort of project I'd be interested in but I had no doubts something would come up that sounded interesting. 
+
+As it happened, I had been playing around with a bit of code, purely out of interest and hoping to learn how [`htmlwidgets`](https://www.htmlwidgets.org/) work. The idea was to make a classic graphic equaliser visualisation like this
+
+![](/img/blog-images/2017-11-10-realtime/graphiceq.gif)
+
+using R.
+
+This presents several challenges; how can I get live audio into R, and how fast can I plot the signal? I had doubts about both parts, partly because of the way that R calls tie up the session ([for now...](https://appsilondatascience.com/blog/rstats/2017/11/01/r-promises-hands-on.html)) and partly because constructing a `ggplot2` object (without starting anything; it would be) is somewhat slow (in terms of raw audio speeds). I'd heard about `htmlwidgets` and thought there must be a way to leverage that towards my goal.
+
+I searched for a graphic equaliser javascript library to work with and didn't find much that aligned with what I had in my head. Eventually I stumbled on [`p5.js`](https://p5js.org/) and its examples page which has an [audio-input plot with a live demo](https://p5js.org/examples/sound-frequency-spectrum.html). It's a frequency spectrum, but I figured that's just a bit of binning away from what I need. Running the example there looks like
+
+![](/img/blog-images/2017-11-10-realtime/p5sound_optimised.gif)
+
+This seemed to be worth a go. I managed to follow enough of [this tutorial](https://www.htmlwidgets.org/develop_intro.html) to have the library called from R. I modified the javascript canvas code to look a little more familiar, and the first iteration of `geom_realtime()` was born.
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/3mdiCUbgxi0" frameborder="0" gesture="media" allowfullscreen></iframe>
+
+This seemed like enough of an idea that I proposed it on the GitHub Issues for the unconf. It got a bit of attention, which was worrying, because I had no idea what to do with this next. Peter Haitch pointed out that Sean Kross had [already wrapped some of the `p5.js` calls into R calls](https://seankross.com/2017/08/11/Beyond-Axes-Simulating-Systems-with-Interactive-Graphics.html) with his `p5` package, so this seemed like a great place to start. It's quite a clever way of doing it too; it involves re-writing the `javascript` which `htmlwidgets` calls on each time you want to do something.
```

Comment:
Peter Haitch should be Peter Hickey. His twitter handle is @PeteHaitch.
You could consider linking Peter and Sean to their GitHub or twitter, whichever most appropriate for each